cranelift/codegen/src/isa/aarch64/lower.rs

```diff
@@ -544,105 +544,241 @@ pub(crate) fn alu_inst_immshift(
-    }
+    // Collect addends through an arbitrary tree of 32-to-64-bit sign/zero
+    // extends and addition ops. We update these as we consume address
+    // components, so they represent the remaining addends not yet handled.
+    let (mut addends64, mut addends32, args_offset) = collect_address_addends(ctx, roots);
+    let mut offset = args_offset + (offset as i64);
 
-    // Handle two regs and a zero offset with built-in extend, if possible.
-    if addends.len() == 2 && offset == 0 {
-        // r1, r2 (to be extended), r2_bits, is_signed
-        let mut parts: Option<(Reg, Reg, usize, bool)> = None;
-        // Handle extension of either first or second addend.
-        for i in 0..2 {
-            if let Some((op, ext_insn)) =
-                maybe_input_insn_multi(ctx, addends[i], &[Opcode::Uextend, Opcode::Sextend])
-            {
-                // Non-extended addend.
-                let r1 = put_input_in_reg(ctx, addends[1 - i], NarrowValueMode::ZeroExtend64);
-                // Extended addend.
-                let r2 = put_input_in_reg(
-                    ctx,
-                    InsnInput {
-                        insn: ext_insn,
-                        input: 0,
-                    },
-                    NarrowValueMode::None,
-                );
-                let r2_bits = ty_bits(ctx.input_ty(ext_insn, 0));
-                parts = Some((
-                    r1,
-                    r2,
-                    r2_bits,
-                    /* is_signed = */ op == Opcode::Sextend,
-                ));
-                break;
-            }
-        }
+    debug!(
+        "lower_address: addends64 {:?}, addends32 {:?}, offset {}",
+        addends64, addends32, offset
+    );
 
-        if let Some((r1, r2, r2_bits, is_signed)) = parts {
-            match (r2_bits, is_signed) {
-                (32, false) => {
-                    return MemArg::RegExtended(r1, r2, ExtendOp::UXTW);
-                }
-                (32, true) => {
-                    return MemArg::RegExtended(r1, r2, ExtendOp::SXTW);
-                }
-                _ => {}
-            }
+    // First, decide what the `MemArg` will be. Take one extendee and one
+    // 64-bit reg, or two 64-bit regs, or one 64-bit reg and an offset.
+    let memarg = if addends32.len() > 0 && addends64.len() > 0 {
+        let (reg32, extendop) = addends32.pop().unwrap();
+        let reg64 = addends64.pop().unwrap();
+        MemArg::RegExtended(reg64, reg32, extendop)
+    } else if offset > 0 && offset < 0x1000 && addends64.len() > 0 {
+        let reg64 = addends64.pop().unwrap();
+        let off = offset;
+        offset = 0;
+        MemArg::RegOffset(reg64, off, elem_ty)
+    } else if addends64.len() >= 2 {
+        let reg1 = addends64.pop().unwrap();
+        let reg2 = addends64.pop().unwrap();
+        MemArg::RegReg(reg1, reg2)
+    } else if addends64.len() == 0 && addends32.len() > 0 {
```

Comment:
Can this case, `addends64.len() == 0 && <whatever>` actually happen?  Wouldn't it imply that whatever created this code knew that it would run in an environment where the max virtual address is 2^32 or some small multiple thereof ?
I guess that's implausible for the general 64-bit target case, but it's not *per se* invalid.

Reply:
Indeed, it's valid CLIF, so we need to either handle it here, or else define it as invalid and punt this problem to the CLIF verifier. It's conceivable that (as you say) the implementer might know better and use 32-bit pointers, so I don't think we can rule it out.

(I actually asserted that this case did not happen at first, but caught myself with a test-case I had previously written that used only a zero-extended 32-bit value as an address... so if nothing else, at least mischevous unit-test authors are known to do this.)